pkg/controllers/user/monitoring/clusterHandler.go

```diff
@@ -368,134 +382,83 @@ func (ch *clusterHandler) deployApp(appName, appTargetNamespace string, appProje
-		checkers = append(checkers, func() error {
-			return isKubeStateExporterWithdrew(ch.app.agentDeploymentClient, appTargetNamespace, appName, monitoringStatus)
-		})
-	}
-	if !ConditionPrometheusDeployed.IsFalse(monitoringStatus) {
-		checkers = append(checkers, func() error {
-			return isPrometheusWithdrew(ch.app.agentStatefulSetClient, appTargetNamespace, appName, monitoringStatus)
-		})
-	}
-	if !ConditionMetricExpressionDeployed.IsFalse(monitoringStatus) {
-		checkers = append(checkers, func() error {
-			return isMetricExpressionWithdrew(cluster.Name, ch.app.cattleClusterGraphClient, ch.app.cattleMonitorMetricClient, monitoringStatus)
-		})
-	}
+	for _, metric := range preDefinedClusterMetrics {
+		newObj := metric.DeepCopy()
+		newObj.Namespace = clusterName
 
-	if len(checkers) == 0 {
-		return nil
+		_, err := ch.app.cattleMonitorMetricClient.Create(newObj)
+		if err != nil && !k8serrors.IsAlreadyExists(err) {
+			return err
+		}
 	}
 
-	err := stream(checkers...)
-	if err != nil {
-		time.Sleep(5 * time.Second)
+	for _, graph := range preDefinedClusterGraph {
+		newObj := graph.DeepCopy()
+		newObj.Namespace = clusterName
+
+		_, err := ch.app.cattleClusterGraphClient.Create(newObj)
+		if err != nil && !k8serrors.IsAlreadyExists(err) {
+			return err
+		}
 	}
 
-	return err
+	return nil
 }
 
-func getClusterTag(cluster *mgmtv3.Cluster) string {
-	return fmt.Sprintf("%s(%s)", cluster.Name, cluster.Spec.DisplayName)
-}
+func (ch *clusterHandler) withdrawMetrics(cluster *mgmtv3.Cluster) error {
+	clusterName := cluster.Name
 
-func isRkeCluster(cluster *mgmtv3.Cluster) bool {
-	return cluster.Status.Driver == mgmtv3.ClusterDriverRKE
-}
+	for _, metric := range preDefinedClusterMetrics {
+		err := ch.app.cattleMonitorMetricClient.DeleteNamespaced(clusterName, metric.Name, &metav1.DeleteOptions{})
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return err
+		}
+	}
 
-func getSecretPath(secretName, name string) string {
-	return fmt.Sprintf("/etc/prometheus/secrets/%s/%s", secretName, name)
-}
+	for _, graph := range preDefinedClusterGraph {
+		err := ch.app.cattleClusterGraphClient.DeleteNamespaced(clusterName, graph.Name, &metav1.DeleteOptions{})
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return err
+		}
+	}
 
-func getCertName(name string) string {
-	return fmt.Sprintf("%s.pem", name)
+	return nil
 }
 
-func getKeyName(name string) string {
-	return fmt.Sprintf("%s-key.pem", name)
+func (ch *clusterHandler) isPrometheusReady(cluster *mgmtv3.Cluster) error {
+	svcName, namespace, _ := monitoring.ClusterPrometheusEndpoint()
+	endpoints, err := ch.agentEndpointsLister.Get(namespace, svcName)
+	if err != nil {
+		return errors.Wrapf(err, "prometheus is not ready")
+	}
+
+	if len(endpoints.Subsets) == 0 || len(endpoints.Subsets[0].Addresses) == 0 {
+		// return ForgetError as clusterMonitoringEnabledHandler will trigger a new loop
+		return &controller.ForgetError{Err: fmt.Errorf("prometheus is not ready")}
```

Comment:
Please don't mix `get an error` and `prometheus nor ready` up as error.

Reply:
Please change to 
```
func (ch *clusterHandler) isPrometheusReady(cluster *mgmtv3.Cluster) (bool,error) 
```

Reply:
It's not necessary to use `controller.ForgetError` here.

Reply:
Updated